tools/pythonpkg/src/pyexpression.cpp

```diff
@@ -0,0 +1,345 @@
+	if (!expression) {
+		throw InternalException("DuckDBPyExpression created without an expression");
+	}
+}
+
+string DuckDBPyExpression::Type() const {
+	return ExpressionTypeToString(expression->type);
+}
+
+string DuckDBPyExpression::ToString() const {
+	return expression->ToString();
+}
+
+void DuckDBPyExpression::Print() const {
+	Printer::Print(expression->ToString());
+}
+
+const ParsedExpression &DuckDBPyExpression::GetExpression() const {
+	return *expression;
+}
+
+shared_ptr<DuckDBPyExpression> DuckDBPyExpression::Copy() const {
+	auto expr = GetExpression().Copy();
+	return make_shared<DuckDBPyExpression>(std::move(expr), order_type, null_order);
+}
+
+shared_ptr<DuckDBPyExpression> DuckDBPyExpression::SetAlias(const string &name) const {
+	auto copied_expression = GetExpression().Copy();
+	copied_expression->alias = name;
+	return make_shared<DuckDBPyExpression>(std::move(copied_expression));
+}
+
+shared_ptr<DuckDBPyExpression> DuckDBPyExpression::Cast(const DuckDBPyType &type) const {
+	auto copied_expression = GetExpression().Copy();
+	auto case_expr = make_uniq<duckdb::CastExpression>(type.Type(), std::move(copied_expression));
+	return make_shared<DuckDBPyExpression>(std::move(case_expr));
+}
+
+// Case Expression modifiers
+
+void DuckDBPyExpression::AssertCaseExpression() const {
+	if (expression->type != ExpressionType::CASE_EXPR) {
+		throw py::value_error("This method can only be used on a Expression resulting from CaseExpression or When");
+	}
+}
+
+shared_ptr<DuckDBPyExpression> DuckDBPyExpression::InternalWhen(unique_ptr<duckdb::CaseExpression> expr,
+                                                                const DuckDBPyExpression &condition,
+                                                                const DuckDBPyExpression &value) {
+	CaseCheck check;
+	check.when_expr = condition.GetExpression().Copy();
+	check.then_expr = value.GetExpression().Copy();
+	expr->case_checks.push_back(std::move(check));
+	return make_shared<DuckDBPyExpression>(std::move(expr));
+}
+
+shared_ptr<DuckDBPyExpression> DuckDBPyExpression::When(const DuckDBPyExpression &condition,
+                                                        const DuckDBPyExpression &value) {
+	AssertCaseExpression();
+	auto expr_p = expression->Copy();
+	auto expr = unique_ptr_cast<ParsedExpression, duckdb::CaseExpression>(std::move(expr_p));
+
+	return InternalWhen(std::move(expr), condition, value);
+}
+
+shared_ptr<DuckDBPyExpression> DuckDBPyExpression::Else(const DuckDBPyExpression &value) {
+	AssertCaseExpression();
+	auto expr_p = expression->Copy();
+	auto expr = unique_ptr_cast<ParsedExpression, duckdb::CaseExpression>(std::move(expr_p));
+
+	expr->else_expr = value.GetExpression().Copy();
+	return make_shared<DuckDBPyExpression>(std::move(expr));
+}
+
+// Binary operators
+
+shared_ptr<DuckDBPyExpression> DuckDBPyExpression::Add(const DuckDBPyExpression &other) {
+	return DuckDBPyExpression::BinaryOperator("+", *this, other);
```

Comment:
Maybe we can have enums for the string function_name argument? Or no?

Reply:
Then we'd have to parse it back to string when we create a FunctionExpression, so I'm not sure if that would be a good idea